kaitai/util.go

```diff
@@ -43,19 +44,23 @@ func ProcessZlib(in []byte) ([]byte, error) {
 	// we could reuse it by using a sync.Pool if this is called in a tight loop.
 	r, err := zlib.NewReader(b)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("ProcessZlib: error initializing zlib reader: %w", err)
 	}
 
-	return ioutil.ReadAll(r)
+	res, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("ProcessZlib: error reading zlib data: %w", err)
+	}
+	return res, nil
 }
 
 // BytesToStr returns a string decoded by the given decoder.
 func BytesToStr(in []byte, decoder *encoding.Decoder) (string, error) {
 	i := bytes.NewReader(in)
 	o := transform.NewReader(i, decoder)
-	d, e := ioutil.ReadAll(o)
-	if e != nil {
-		return "", e
+	d, err := ioutil.ReadAll(o)
+	if err != nil {
+		return "", fmt.Errorf("BytesToStr: error reading all: %w", err)
```

Comment:
I don't like the `reading all` in the message, it doesn't capture the essence of the error. I suggest something like `error decoding bytes`, and it turns out we can sometimes reveal which encoding was used by printing the type of `decoder.Transformer`, so we can include it in the message too:

```suggestion
		return "", fmt.Errorf("BytesToStr: error decoding bytes with %T: %w", decoder.Transformer, err)
```

For example (https://go.dev/play/p/Ljcbzu8EnSw):

```go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"

	"golang.org/x/text/encoding"
	"golang.org/x/text/encoding/unicode"
	"golang.org/x/text/transform"
)

func BytesToStr(in []byte, decoder *encoding.Decoder) (string, error) {
	i := bytes.NewReader(in)
	o := transform.NewReader(i, decoder)
	d, err := ioutil.ReadAll(o)
	if err != nil {
		return "", fmt.Errorf("BytesToStr: error decoding bytes with %T: %w", decoder.Transformer, err)
	}
	return string(d), nil
}

func main() {
	dec := unicode.UTF16(unicode.BigEndian, unicode.ExpectBOM).NewDecoder()
	s, err := BytesToStr([]uint8{ /* 0xfe, 0xff, */ 0x30, 0x53, 0x30, 0x93}, dec)
	fmt.Printf("%#v\n", s)
	fmt.Printf("%v\n", err)
}
```

Output:

```
""
BytesToStr: error decoding bytes with *unicode.utf16Decoder: encoding: missing byte order mark
```

(The type of `decoder.Transformer` isn't _always_ useful - for example, for `charmap.CodePage437` from `"golang.org/x/text/encoding/charmap"`, it prints only `charmap.charmapDecoder`. On the other hand, I'm not sure whether the `charmap` encodings can fail, they seem to solve any issue with the U+FFFD replacement character.)

Reply:
Sounds good, let me commit this separately!